### PR TITLE
Implement credential_arn for ecr pull cache through rules

### LINKED ISSUE #34466: [Enhancement]: aws_ecr_pull_through_cache_rule support CredentialArn and UpstreamRegistry
### Description

AWS ECR Pull Through Cache Rules now support two additional parameters: upstreamRegistry and credentialArn 

These are required to support pulling from (for example) Docker Hub and other  types of authenticated upstreams.

### Affected Resource(s) and/or Data Source(s)

aws_ecr_pull_through_cache_rule

### Potential Terraform Configuration

```terraform
resource "aws_ecr_pull_through_cache_rule" "docker-hub" {
  ecr_repository_prefix = "docker-hub"
  upstream_registry_url = "registry-1.docker.io"
  credential_arn = "arn:aws:secretsmanager:us-east-2:111122223333:secret:ecr-pullthroughcache/example1234"
}
```

### References

https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_CreatePullThroughCacheRule.html
https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-through-cache-creating-rule.html

### Would you like to implement a fix?

No

### PR DESCRIPTION
**Pull request by Baachi**
### Description
As mentioned in #34466, AWS released the support for authenticated docker registries in ECR. This MR adds the new credentialArn to the terraform datasource and resource. 

### Relations

Closes #34466

### References

https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_CreatePullThroughCacheRule.html
https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-through-cache-creating-rule.html

### Output from Acceptance Testing

```console
% make testacc TESTS=TestAccECRPullThroughCacheRule_credentialArn PKG=ecr
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecr/... -v -count 1 -parallel 20 -run='TestAccECRPullThroughCacheRule_credentialArn'  -timeout 360m
=== RUN   TestAccECRPullThroughCacheRule_credentialArn
=== PAUSE TestAccECRPullThroughCacheRule_credentialArn
=== CONT  TestAccECRPullThroughCacheRule_credentialArn
--- PASS: TestAccECRPullThroughCacheRule_credentialArn (30.69s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ecr	32.989s


% make testacc TESTS=TestAccECRPullThroughCacheRuleDataSource_credential PKG=ecr
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecr/... -v -count 1 -parallel 20 -run='TestAccECRPullThroughCacheRuleDataSource_credential'  -timeout 360m
=== RUN   TestAccECRPullThroughCacheRuleDataSource_credential
=== PAUSE TestAccECRPullThroughCacheRuleDataSource_credential
=== CONT  TestAccECRPullThroughCacheRuleDataSource_credential
--- PASS: TestAccECRPullThroughCacheRuleDataSource_credential (27.80s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ecr	30.1
```

### CODE REVIEW

**Comment by Baachi** [comment]
MR is ready for review!

**Comment by AdamTylerLynch** [comment]
Related: #34716

File: .changelog/34475.txt
Comment (AdamTylerLynch) [inline .changelog/34475.txt:2]: Please include documentation updates:

1. /website/docs/r/ecr_pull_through_cache_rule.html.markdown
2. /website/docs/d/ecr_pull_through_cache_rule.html.markdown

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
This is a wonderful PR! Thank you so much for your contribution! I'm requesting a few small additions (documentation for the resource and datasource) and some small linting modifications.

Please include documentation updates:

1. /website/docs/r/ecr_pull_through_cache_rule.html.markdown
2. /website/docs/d/ecr_pull_through_cache_rule.html.markdown

**Comment by Baachi** [comment]
Hey @AdamTylerLynch , 

thanks for the review!  I fixed all your comments (or I hope at least.) 😁

File: internal/service/ecr/pull_through_cache_rule.go
Comment (thefirstofthe300) [inline internal/service/ecr/pull_through_cache_rule.go:58]: The `:aws:` portion of the ARN can change depending on if you're using the GovCloud or China regions. It probably would behoove us to not require that the normal `aws` partition be part of the ARN.

Comment (Baachi) [reply internal/service/ecr/pull_through_cache_rule.go:58]: @thefirstofthe300 Yeah I already notice it. This is just the pattern written in the documentation (and in the API error message). I can change it, but then users will get the error message from AWS, not Terraform. Is it the best way?

https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-pullthroughcacherule.html

Comment (blakepettersson) [reply internal/service/ecr/pull_through_cache_rule.go:58]: What about this? This would fail if it doesn't match the existing partitions (aws, aws-cn, aws-us-gov)

```suggestion
					regexache.MustCompile(`^arn:aws(\-cn|\-us\-gov)?:secretsmanager:[a-zA-Z0-9-:]+:secret:ecr\-pullthroughcache\/[a-zA-Z0-9\/_+=.@-]+$`),
```

Comment (thefirstofthe300) [reply internal/service/ecr/pull_through_cache_rule.go:58]: You raise a good point. Two china regions and the govcloud regions don't appear to be supported. For now, it's probably better to leave as you have it.

**Comment by Baachi** [comment]
Any updates here?

**Comment by blakepettersson** [comment]
@Baachi seems like that there are some merge conflicts now

File: internal/service/ecr/pull_through_cache_rule.go
Comment (AdamTylerLynch) [inline internal/service/ecr/pull_through_cache_rule.go:57]: There are multiple AWS partitions, recommend you simply validate that a ARN is provided, and let the service API do the rest.

```suggestion
				ValidateFunc: verify.ValidARN,
```

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
Last requested change

File: internal/service/ecr/pull_through_cache_rule.go
Comment (Baachi) [reply internal/service/ecr/pull_through_cache_rule.go]: Done ✅

**Comment by Baachi** [comment]
@AdamTylerLynch fixed your comment and rebased against master :)

**Comment by alileza** [comment]
Let's merge this 🚀

**Comment by honarkhah** [comment]
Any update?

**Comment by sebastienalbert** [comment]
Hi,

Do you have any idea when this feature will be available?

**Comment by Baachi** [comment]
Hey @AdamTylerLynch can I do something here to get this merged/reviewed?

**Review by honarkhah — APPROVED** [review]

**Comment by AndreZiviani** [comment]
Any progress here? :pray:

**Comment by sharovmerk** [comment]
Hi, when it can be merged?

**Comment by sharovmerk** [comment]
btw. don't you know, if it possible to create `aws_ecr_pull_through_cache_rule` for dockerhub without credentials?

**Comment by kvendingoldo** [comment]
btw. can you add an example of secret_string for aws_secretsmanager_secret_version?

**Comment by eugenetaranov** [comment]
> btw. don't you know, if it possible to create `aws_ecr_pull_through_cache_rule` for dockerhub without credentials?

afaik no, dockerhub now requires everyone to be authenticated

**Comment by kvendingoldo** [comment]
> afaik no, dockerhub now requires everyone to be authenticated
Yep, I have already checked docs. It's a little bit painful that I users can't use proxy without auth for public repositories.

**Comment by sharovmerk** [comment]
@ewbankkit can you check the PR, pls? It's really important feature for the production.

**Comment by ben-bertrands-hs** [comment]
@Baachi I think you're good to merge this in

**Comment by ben-bertrands-hs** [comment]
@honarkhah can you help merge this in?

**Comment by Baachi** [comment]
Im not a maintainer, so I can't merge it :)

**Comment by ben-bertrands-hs** [comment]
> Im not a maintainer, so I can't merge it :)

Yeah sorry, I was a bit to quick commenting there 😅

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccECRPullThroughCacheRuleDataSource_\|TestAccECRPullThroughCacheRule_' PKG=ecr ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecr/... -v -count 1 -parallel 3  -run=TestAccECRPullThroughCacheRuleDataSource_\|TestAccECRPullThroughCacheRule_ -timeout 360m
=== RUN   TestAccECRPullThroughCacheRuleDataSource_basic
=== PAUSE TestAccECRPullThroughCacheRuleDataSource_basic
=== RUN   TestAccECRPullThroughCacheRuleDataSource_repositoryPrefixWithSlash
=== PAUSE TestAccECRPullThroughCacheRuleDataSource_repositoryPrefixWithSlash
=== RUN   TestAccECRPullThroughCacheRuleDataSource_credential
=== PAUSE TestAccECRPullThroughCacheRuleDataSource_credential
=== RUN   TestAccECRPullThroughCacheRule_basic
=== PAUSE TestAccECRPullThroughCacheRule_basic
=== RUN   TestAccECRPullThroughCacheRule_credentialARN
=== PAUSE TestAccECRPullThroughCacheRule_credentialARN
=== RUN   TestAccECRPullThroughCacheRule_disappears
=== PAUSE TestAccECRPullThroughCacheRule_disappears
=== RUN   TestAccECRPullThroughCacheRule_failWhenAlreadyExists
=== PAUSE TestAccECRPullThroughCacheRule_failWhenAlreadyExists
=== RUN   TestAccECRPullThroughCacheRule_repositoryPrefixWithSlash
=== PAUSE TestAccECRPullThroughCacheRule_repositoryPrefixWithSlash
=== CONT  TestAccECRPullThroughCacheRuleDataSource_basic
=== CONT  TestAccECRPullThroughCacheRule_credentialARN
=== CONT  TestAccECRPullThroughCacheRule_failWhenAlreadyExists
--- PASS: TestAccECRPullThroughCacheRule_failWhenAlreadyExists (25.43s)
=== CONT  TestAccECRPullThroughCacheRuleDataSource_credential
--- PASS: TestAccECRPullThroughCacheRuleDataSource_basic (36.58s)
=== CONT  TestAccECRPullThroughCacheRule_basic
--- PASS: TestAccECRPullThroughCacheRule_credentialARN (41.44s)
=== CONT  TestAccECRPullThroughCacheRuleDataSource_repositoryPrefixWithSlash
--- PASS: TestAccECRPullThroughCacheRuleDataSource_credential (23.19s)
=== CONT  TestAccECRPullThroughCacheRule_disappears
--- PASS: TestAccECRPullThroughCacheRule_basic (21.31s)
=== CONT  TestAccECRPullThroughCacheRule_repositoryPrefixWithSlash
--- PASS: TestAccECRPullThroughCacheRuleDataSource_repositoryPrefixWithSlash (17.94s)
--- PASS: TestAccECRPullThroughCacheRule_disappears (17.97s)
--- PASS: TestAccECRPullThroughCacheRule_repositoryPrefixWithSlash (16.18s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ecr	85.729s
```

**Comment by ewbankkit** [comment]
@Baachi Thanks for the contribution :tada: :clap:.

**Pull request merged**
